File: backend_app/app/models/permissions.py
from enum import Enum
from typing import Dict, Any
from typing import List

class PermissionLevel(str, Enum):
    """Permission levels in hierarchical order"""
    USER = "User"
    EDITOR = "Editor"
    ADMIN = "Admin"

# Permission hierarchy (higher number = more permissions)
PERMISSION_HIERARCHY = {
    PermissionLevel.USER.value: 1,    # "User": 1
    PermissionLevel.EDITOR.value: 2,  # "Editor": 2
    PermissionLevel.ADMIN.value: 3,   # "Admin": 3
}

def get_permission_level(permission_string: str) -> int:
    """
    Get the numeric level for a permission string.
    
    Args:
        permission_string: The permission level as a string (e.g., 'Admin', 'Editor', 'User')
    
    Returns:
        int: The numeric permission level (1-3), or 0 if invalid
    """
    return PERMISSION_HIERARCHY.get(permission_string, 0)

def has_permission_level(user_permission: str, required_permission: str) -> bool:
    """
    Check if a user has the required permission level or higher.
    
    Args:
        user_permission: The user's permission level string
        required_permission: The required permission level string
    
    Returns:
        bool: True if user has sufficient permissions, False otherwise
    """
    user_level = get_permission_level(user_permission)
    required_level = get_permission_level(required_permission)
    return user_level >= required_level


# Default capability lists (keep concise; frontend files enumerate expected keys)
DEFAULT_CAPABILITIES: Dict[str, List[str]] = {
    PermissionLevel.USER.value: [
        "can_view_prompts",
        "can_create_prompts",
        "can_upload_files",
    ],
    PermissionLevel.EDITOR.value: [
        "can_view_prompts",
        "can_create_prompts",
        "can_edit_prompts",
        "can_upload_files",
        "can_download_files",
    ],
    PermissionLevel.ADMIN.value: [
        # Admin gets every known capability; callers may convert this to full map
    ],
}


def capabilities_for_permission(permission: str, all_capabilities: List[str]) -> Dict[str, bool]:
    """Return a mapping of capability -> bool for a given permission level.

    - Admin: all capabilities True
    - Editor: defaults from DEFAULT_CAPABILITIES[Editor]
    - User: defaults from DEFAULT_CAPABILITIES[User]
    """
    if permission == PermissionLevel.ADMIN.value:
        return {cap: True for cap in all_capabilities}

    enabled = set(DEFAULT_CAPABILITIES.get(permission, []))
    return {cap: (cap in enabled) for cap in all_capabilities}


# --- Backwards compatibility: PermissionCapability enum and helpers ---
class PermissionCapability(str, Enum):
    CAN_VIEW_OWN_JOBS = "can_view_own_jobs"
    CAN_CREATE_JOBS = "can_create_jobs"
    CAN_EDIT_OWN_JOBS = "can_edit_own_jobs"
    CAN_DELETE_OWN_JOBS = "can_delete_own_jobs"
    CAN_VIEW_SHARED_JOBS = "can_view_shared_jobs"
    CAN_EDIT_SHARED_JOBS = "can_edit_shared_jobs"
    CAN_DELETE_SHARED_JOBS = "can_delete_shared_jobs"
    CAN_SHARE_JOBS = "can_share_jobs"
    CAN_VIEW_ALL_JOBS = "can_view_all_jobs"
    CAN_EDIT_ALL_JOBS = "can_edit_all_jobs"
    CAN_DELETE_ALL_JOBS = "can_delete_all_jobs"
    CAN_VIEW_PROMPTS = "can_view_prompts"
    CAN_CREATE_PROMPTS = "can_create_prompts"
    CAN_EDIT_PROMPTS = "can_edit_prompts"
    CAN_DELETE_PROMPTS = "can_delete_prompts"
    CAN_CREATE_TEMPLATES = "can_create_templates"
    CAN_VIEW_USERS = "can_view_users"
    CAN_CREATE_USERS = "can_create_users"
    CAN_EDIT_USERS = "can_edit_users"
    CAN_DELETE_USERS = "can_delete_users"
    CAN_MANAGE_USERS = "can_manage_users"
    CAN_VIEW_SETTINGS = "can_view_settings"
    CAN_EDIT_SETTINGS = "can_edit_settings"
    CAN_VIEW_ANALYTICS = "can_view_analytics"
    CAN_MANAGE_SYSTEM = "can_manage_system"
    CAN_UPLOAD_FILES = "can_upload_files"
    CAN_DOWNLOAD_FILES = "can_download_files"
    CAN_EXPORT_DATA = "can_export_data"
    CAN_IMPORT_DATA = "can_import_data"


def get_user_capabilities(user_permission: str, custom_capabilities: Dict[str, Any]) -> Dict[str, bool]:
    """Return effective capability mapping (base + custom) for a user."""
    # Build all capability keys from PermissionCapability
    all_caps = [c.value for c in PermissionCapability]
    base = capabilities_for_permission(user_permission, all_caps)
    # Merge custom capabilities (override)
    for k, v in (custom_capabilities or {}).items():
        if k in base:
            base[k] = bool(v)
        else:
            base[k] = bool(v)
    return base


def merge_custom_capabilities(base_caps: Dict[str, bool], custom_caps: Dict[str, Any]) -> Dict[str, bool]:
    """Merge custom capability flags into base capability map."""
    merged = base_caps.copy()
    for k, v in (custom_caps or {}).items():
        merged[k] = bool(v)
    return merged

File: frontend_app/src/types/permissions.ts
// Resource-based permission system for Sonic Brief - matches backend

export enum PermissionLevel {
  USER = "User",
  EDITOR = "Editor", 
  ADMIN = "Admin"
}

export enum Capability {
  // Job/Transcription Management
  CAN_VIEW_OWN_JOBS = "can_view_own_jobs",
  CAN_CREATE_JOBS = "can_create_jobs",
  CAN_EDIT_OWN_JOBS = "can_edit_own_jobs",
  CAN_DELETE_OWN_JOBS = "can_delete_own_jobs",
  CAN_VIEW_SHARED_JOBS = "can_view_shared_jobs",
  CAN_EDIT_SHARED_JOBS = "can_edit_shared_jobs",
  CAN_DELETE_SHARED_JOBS = "can_delete_shared_jobs",
  CAN_SHARE_JOBS = "can_share_jobs",
  CAN_VIEW_ALL_JOBS = "can_view_all_jobs",
  CAN_EDIT_ALL_JOBS = "can_edit_all_jobs",
  CAN_DELETE_ALL_JOBS = "can_delete_all_jobs",

  // Prompt/Template Management
  CAN_VIEW_PROMPTS = "can_view_prompts",
  CAN_CREATE_PROMPTS = "can_create_prompts",
  CAN_EDIT_PROMPTS = "can_edit_prompts",
  CAN_DELETE_PROMPTS = "can_delete_prompts",
  CAN_CREATE_TEMPLATES = "can_create_templates",

  // User Management
  CAN_VIEW_USERS = "can_view_users",
  CAN_CREATE_USERS = "can_create_users",
  CAN_EDIT_USERS = "can_edit_users",
  CAN_DELETE_USERS = "can_delete_users",
  CAN_MANAGE_USERS = "can_manage_users",

  // System/Settings
  CAN_VIEW_SETTINGS = "can_view_settings",
  CAN_EDIT_SETTINGS = "can_edit_settings",
  CAN_VIEW_ANALYTICS = "can_view_analytics",
  CAN_MANAGE_SYSTEM = "can_manage_system",

  // File Operations
  CAN_UPLOAD_FILES = "can_upload_files",
  CAN_DOWNLOAD_FILES = "can_download_files",
  CAN_EXPORT_DATA = "can_export_data",
  CAN_IMPORT_DATA = "can_import_data",

  // Legacy capabilities for backward compatibility
  VIEW_TRANSCRIPTIONS = "can_view_own_jobs",
  CREATE_TRANSCRIPTIONS = "can_create_jobs",
  EDIT_TRANSCRIPTIONS = "can_edit_own_jobs",
  DELETE_TRANSCRIPTIONS = "can_delete_own_jobs",
  VIEW_USERS = "can_view_users",
  CREATE_USERS = "can_create_users",
  EDIT_USERS = "can_edit_users",
  DELETE_USERS = "can_delete_users",
  VIEW_SETTINGS = "can_view_settings",
  EDIT_SETTINGS = "can_edit_settings",
  MANAGE_SYSTEM = "can_manage_system",
  VIEW_ANALYTICS = "can_view_analytics",
  EXPORT_DATA = "can_export_data",
  IMPORT_DATA = "can_import_data"
}

export interface UserCapabilities {
  [key: string]: boolean;
}

// Permission hierarchy (higher number = more permissions)
export const PERMISSION_HIERARCHY: Record<PermissionLevel, number> = {
  [PermissionLevel.USER]: 1,
  [PermissionLevel.EDITOR]: 2,
  [PermissionLevel.ADMIN]: 3,
};

// Default capabilities for each permission level (matches backend)
export const DEFAULT_CAPABILITIES: Record<PermissionLevel, Capability[]> = {
  [PermissionLevel.USER]: [
    Capability.CAN_VIEW_OWN_JOBS,
    Capability.CAN_CREATE_JOBS,
    Capability.CAN_EDIT_OWN_JOBS,
    Capability.CAN_DELETE_OWN_JOBS,
    Capability.CAN_VIEW_SHARED_JOBS,
    Capability.CAN_VIEW_PROMPTS,
    Capability.CAN_VIEW_SETTINGS,
    Capability.CAN_UPLOAD_FILES,
    Capability.CAN_DOWNLOAD_FILES,
  ],
  [PermissionLevel.EDITOR]: [
    Capability.CAN_VIEW_OWN_JOBS,
    Capability.CAN_CREATE_JOBS,
    Capability.CAN_EDIT_OWN_JOBS,
    Capability.CAN_DELETE_OWN_JOBS,
    Capability.CAN_VIEW_SHARED_JOBS,
    Capability.CAN_EDIT_SHARED_JOBS,
    Capability.CAN_DELETE_SHARED_JOBS,
    Capability.CAN_SHARE_JOBS,
    Capability.CAN_VIEW_PROMPTS,
    Capability.CAN_CREATE_PROMPTS,
    Capability.CAN_EDIT_PROMPTS,
    Capability.CAN_CREATE_TEMPLATES,
    Capability.CAN_VIEW_SETTINGS,
    Capability.CAN_EDIT_SETTINGS,
    Capability.CAN_UPLOAD_FILES,
    Capability.CAN_DOWNLOAD_FILES,
    Capability.CAN_EXPORT_DATA,
  ],
  [PermissionLevel.ADMIN]: [
    // All capabilities enabled for admin
    Capability.CAN_VIEW_OWN_JOBS,
    Capability.CAN_CREATE_JOBS,
    Capability.CAN_EDIT_OWN_JOBS,
    Capability.CAN_DELETE_OWN_JOBS,
    Capability.CAN_VIEW_SHARED_JOBS,
    Capability.CAN_EDIT_SHARED_JOBS,
    Capability.CAN_DELETE_SHARED_JOBS,
    Capability.CAN_SHARE_JOBS,
    Capability.CAN_VIEW_ALL_JOBS,
    Capability.CAN_EDIT_ALL_JOBS,
    Capability.CAN_DELETE_ALL_JOBS,
    Capability.CAN_VIEW_PROMPTS,
    Capability.CAN_CREATE_PROMPTS,
    Capability.CAN_EDIT_PROMPTS,
    Capability.CAN_DELETE_PROMPTS,
    Capability.CAN_CREATE_TEMPLATES,
    Capability.CAN_VIEW_USERS,
    Capability.CAN_CREATE_USERS,
    Capability.CAN_EDIT_USERS,
    Capability.CAN_DELETE_USERS,
    Capability.CAN_MANAGE_USERS,
    Capability.CAN_VIEW_SETTINGS,
    Capability.CAN_EDIT_SETTINGS,
    Capability.CAN_VIEW_ANALYTICS,
    Capability.CAN_MANAGE_SYSTEM,
    Capability.CAN_UPLOAD_FILES,
    Capability.CAN_DOWNLOAD_FILES,
    Capability.CAN_EXPORT_DATA,
    Capability.CAN_IMPORT_DATA,
  ]
};

// Grouped capabilities for UI display
export const CAPABILITY_GROUPS = {
  jobs: {
    label: "Job Management",
    capabilities: [
      Capability.CAN_VIEW_OWN_JOBS,
      Capability.CAN_CREATE_JOBS,
      Capability.CAN_EDIT_OWN_JOBS,
      Capability.CAN_DELETE_OWN_JOBS,
      Capability.CAN_VIEW_SHARED_JOBS,
      Capability.CAN_EDIT_SHARED_JOBS,
      Capability.CAN_DELETE_SHARED_JOBS,
      Capability.CAN_SHARE_JOBS,
      Capability.CAN_VIEW_ALL_JOBS,
      Capability.CAN_EDIT_ALL_JOBS,
      Capability.CAN_DELETE_ALL_JOBS,
    ]
  },
  prompts: {
    label: "Prompt Management", 
    capabilities: [
      Capability.CAN_VIEW_PROMPTS,
      Capability.CAN_CREATE_PROMPTS,
      Capability.CAN_EDIT_PROMPTS,
      Capability.CAN_DELETE_PROMPTS,
      Capability.CAN_CREATE_TEMPLATES,
    ]
  },
  users: {
    label: "User Management", 
    capabilities: [
      Capability.CAN_VIEW_USERS,
      Capability.CAN_CREATE_USERS,
      Capability.CAN_EDIT_USERS,
      Capability.CAN_DELETE_USERS,
      Capability.CAN_MANAGE_USERS,
    ]
  },
  settings: {
    label: "Settings",
    capabilities: [
      Capability.CAN_VIEW_SETTINGS,
      Capability.CAN_EDIT_SETTINGS,
      Capability.CAN_VIEW_ANALYTICS,
      Capability.CAN_MANAGE_SYSTEM,
    ]
  },
  files: {
    label: "File Operations",
    capabilities: [
      Capability.CAN_UPLOAD_FILES,
      Capability.CAN_DOWNLOAD_FILES,
      Capability.CAN_EXPORT_DATA,
      Capability.CAN_IMPORT_DATA,
    ]
  }
};

// Utility functions for permission checks
export function hasCapability(userCapabilities: UserCapabilities, capability: Capability): boolean {
  return userCapabilities[capability] === true;
}

export function hasAnyCapability(userCapabilities: UserCapabilities, capabilities: Capability[]): boolean {
  return capabilities.some(cap => hasCapability(userCapabilities, cap));
}

export function hasAllCapabilities(userCapabilities: UserCapabilities, capabilities: Capability[]): boolean {
  return capabilities.every(cap => hasCapability(userCapabilities, cap));
}

// Convert between backend array format and frontend object format
export function capabilitiesToArray(userCapabilities: UserCapabilities): string[] {
  return Object.entries(userCapabilities)
    .filter(([_, enabled]) => enabled)
    .map(([capability, _]) => capability);
}

export function capabilitiesFromArray(capabilities: string[]): UserCapabilities {
  const result: UserCapabilities = {};
  
  // Initialize all capabilities to false
  Object.values(Capability).forEach(cap => {
    result[cap] = false;
  });
  
  // Set enabled capabilities to true
  capabilities.forEach(cap => {
    if (Object.values(Capability).includes(cap as Capability)) {
      result[cap] = true;
    }
  });
  
  return result;
}

// Get user capabilities for a permission level
export function getUserCapabilities(level: PermissionLevel): UserCapabilities {
  const capabilities = DEFAULT_CAPABILITIES[level];
  return capabilitiesFromArray(capabilities);
}

// Get permission level from capabilities
export function getPermissionLevel(userCapabilities: UserCapabilities): PermissionLevel {
  const enabledCapabilities = capabilitiesToArray(userCapabilities) as Capability[];
  
  // Check if user has all admin capabilities
  const adminCapabilities = DEFAULT_CAPABILITIES[PermissionLevel.ADMIN];
  if (adminCapabilities.every(cap => enabledCapabilities.includes(cap))) {
    return PermissionLevel.ADMIN;
  }
  
  // Check if user has all editor capabilities  
  const editorCapabilities = DEFAULT_CAPABILITIES[PermissionLevel.EDITOR];
  if (editorCapabilities.every(cap => enabledCapabilities.includes(cap))) {
    return PermissionLevel.EDITOR;
  }
  
  // Default to user level
  return PermissionLevel.USER;
}

// Additional utility functions
export function hasPermissionLevel(userPermission: PermissionLevel, requiredPermission: PermissionLevel): boolean {
  const userLevel = PERMISSION_HIERARCHY[userPermission] || 0;
  const requiredLevel = PERMISSION_HIERARCHY[requiredPermission] || 0;
  return userLevel >= requiredLevel;
}

export function arrayToCapabilities(capabilities: string[]): UserCapabilities {
  return capabilitiesFromArray(capabilities);
}

export function getCapabilitiesForPermission(permission: PermissionLevel): UserCapabilities {
  const capabilities = DEFAULT_CAPABILITIES[permission] || [];
  return capabilitiesFromArray(capabilities);
}

// Type for user data from API
export interface User {
  id: string;
  email: string;
  permission: PermissionLevel;
  capabilities?: string[];
  custom_capabilities?: UserCapabilities;
  created_at: string;
  updated_at?: string;
}

// Type for permission update requests
export interface PermissionUpdateRequest {
  permission?: PermissionLevel;
  custom_capabilities?: UserCapabilities;
}
